core/src/main/java/io/grpc/ServerServiceDefinition.java

```diff
@@ -102,22 +105,97 @@ private Builder(String serviceName) {
     }
 
     /** Add a method to be supported by the service. */
-    @ExperimentalApi("https://github.com/grpc/grpc-java/issues/1774")
     public <ReqT, RespT> Builder addMethod(ServerMethodDefinition<ReqT, RespT> def) {
       MethodDescriptor<ReqT, RespT> method = def.getMethodDescriptor();
       checkArgument(
-          serviceName.equals(MethodDescriptor.extractFullServiceName(method.getFullMethodName())),
+          serviceDescriptor.getName().equals(
+              MethodDescriptor.extractFullServiceName(method.getFullMethodName())),
           "Service name mismatch. Expected service name: '%s'. Actual method name: '%s'.",
-          this.serviceName, method.getFullMethodName());
+          serviceDescriptor.getName(), method.getFullMethodName());
       String name = method.getFullMethodName();
       checkState(!methods.containsKey(name), "Method by same name already registered: %s", name);
       methods.put(name, def);
       return this;
     }
 
-    /** Construct new ServerServiceDefinition. */
+    /**
+     * Construct new ServerServiceDefinition.
+     */
     public ServerServiceDefinition build() {
-      return new ServerServiceDefinition(serviceName, methods);
+      Map<String, ServerMethodDefinition<?, ?>> tmpMethods =
+          new HashMap<String, ServerMethodDefinition<?, ?>>(methods);
+      for (MethodDescriptor<?, ?> descriptorMethod : serviceDescriptor.getMethods()) {
+        ServerMethodDefinition<?, ?> removed = tmpMethods.remove(
+            descriptorMethod.getFullMethodName());
```

Comment:
Hard equality is checked below, better to still use names for approximate equivalence and better error reporting